File: instar.R
library(TMB)
library(gulf.data)
library(gulf.spatial)
library(gulf.graphics)
library(gulf.stats)

years  <- 2017
sex <- 2
step <- 0.5
if (sex == 1){
   n_instar <- 9
   xlim <- c(0, 140)
   ylim <- c(0, 40)
}else{
   n_instar <- 7
   xlim <- c(0, 100)
   ylim <- c(0, 40)   
}

source("instar.data.R")
source("instar.plot.R")

# Work computer fix:
if (Sys.getenv("RSTUDIO_USER_IDENTITY") == "SuretteTJ") Sys.setenv(BINPREF = "C:/Rtools/mingw_64/bin/")

compile("instar.cpp")
dyn.load(dynlib("instar"))

# Define initial parameters:
parameters <- list(mu0                     = 10,                       # First instar mean size.
                   log_sigma0              = log(0.8),                 # Log-scale standard error for first instar.
                   log_hiatt_slope         = log(c(0.350, 0.0920)),    # Hiatt slope parameters.
                   log_hiatt_intercept     = log(c(0.689, 8.000)),     # Hiatt intercept parameters.
                   log_growth_error        = log(c(0.01, 0.10)),       # Growth increment error inflation parameters
                   log_lambda_alpha        = 7.5,                      # Log-scale global mean density.
                   log_lambda_instar       = rep(0, n_instar),         # Log-scale instar effect.
                   log_sigma_lambda_instar = -1,                       # Log-scale instar effect error parameter.
                   logit_scale             = rep(0, n_instar))

# Fit instar abundances:
map <- lapply(parameters, function(x) factor(rep(NA, length(x))))
map$log_lambda_alpha  <- factor(1)
map$log_lambda_instar <- factor(1:length(parameters$log_lambda_instar))
map$logit_scale       <- factor(1:length(parameters$logit_scale))
map$log_sigma_lambda_instar <- factor(1)
obj <- MakeADFun(data, parameters, DLL = "instar_count", map = map, random = "log_lambda_instar") 
theta <- optim(obj$par, obj$fn, control = list(trace = 3, maxit = 1000))$par
obj$par <- theta
parameters$log_lambda_alpha <- theta[["log_lambda_alpha"]]
parameters$log_sigma_lambda_instar <- theta[["log_sigma_lambda_instar"]]
parameters$log_lambda_instar <- obj$report()$log_lambda_instar
parameters$logit_scale <- obj$report()$logit_scale

# Add error parameters:
map$log_sigma0 = factor(1)
map$log_growth_error = factor(c(1, 2))
obj <- MakeADFun(data, parameters, DLL = "instar_count", map = map, random = "log_lambda_instar") 
theta <- optim(obj$par, obj$fn, control = list(trace = 3, maxit = 500))$par
obj$par <- theta
parameters$log_lambda_alpha <- theta[["log_lambda_alpha"]]
parameters$log_sigma_lambda_instar <- theta[["log_sigma_lambda_instar"]]
parameters$log_lambda_instar <- obj$report()$log_lambda_instar
parameters$logit_scale <- obj$report()$logit_scale
parameters$log_growth_error <- theta[grep("log_growth_error", names(theta))]
parameters$log_sigma0 <- theta[grep("log_sigma0", names(theta))]

obj <- MakeADFun(data, parameters, DLL = "instar_count", random = "log_lambda_instar") 
theta <- optim(obj$par, obj$fn, control = list(trace = 3, maxit = 5000))$par
obj$par <- theta

plot.instar(obj, data, xlim = xlim, ylim = c(0, 180))

100 * obj$report()$scale

File: instar.cpp
#include <TMB.hpp>
template<class Type> Type objective_function<Type>::operator()(){
   // Data:
   DATA_VECTOR(x_imm);          // Immature size measurements (ni).
   DATA_VECTOR(f_imm);          // Immature frequency observations (ni).
   DATA_VECTOR(x_mat);          // Mature size measurements (nm).
   DATA_VECTOR(f_mat);          // Mature frequency observations (nm).
   DATA_SCALAR(dx);             // Size-bin width.   
   
   // Instar growth parameters:                       
   PARAMETER(mu0);                          // First instar mean size.
   PARAMETER(log_sigma0);                   // Log-scale standard error for first instar.
   PARAMETER_VECTOR(log_hiatt_slope);       // Hiatt slope parameters.
   PARAMETER_VECTOR(log_hiatt_intercept);   // Hiatt intercept parameters.
   PARAMETER_VECTOR(log_growth_error);      // Growth increment error inflation parameters.
   
   // Density parameters:
   PARAMETER(log_lambda_alpha);             // Log-scale global mean density.
   PARAMETER_VECTOR(log_lambda_instar);     // Log-scale instar mean density.
   PARAMETER(log_sigma_lambda_instar);      // Log-scale instar mean density error parameter.
   PARAMETER_VECTOR(logit_scale);           // Logit-scale factors.
   
   // Vector sizes:      
   int n_imm = x_imm.size();                // Number of immature observations.
   int n_mat = x_mat.size();                // Number of mature observations.
   int n_instar = log_lambda_instar.size(); // Number instars in model.
   
   // Calculate instar means and errors:
   vector<Type> mu(n_instar+1);
   vector<Type> log_sigma(n_instar+1);
   mu[0] = mu0;
   log_sigma[0] = log_sigma0; 
   for (int k = 1; k < 5; k++){
      mu[k] = exp(log_hiatt_intercept[0]) + mu[k-1] + exp(log_hiatt_slope[0]) * mu[k-1];
      log_sigma[k] = log(1 + exp(log_hiatt_slope[0]) + exp(log_growth_error[0])) + log_sigma[k-1];
   }
   for (int k = 5; k < (n_instar+1); k++){
      mu[k] = exp(log_hiatt_intercept[1]) + mu[k-1] + exp(log_hiatt_slope[1]) * mu[k-1];
      log_sigma[k] = log(1 + exp(log_hiatt_slope[1]) + exp(log_growth_error[1])) + log_sigma[k-1];
   }
   vector<Type> sigma = exp(log_sigma);
   
   // Initialize log-likelihood:
   Type v = 0;
   
   // Random effects:
   v -= sum(dnorm(log_lambda_instar, 0, exp(log_sigma_lambda_instar), true));  // Instar-level random effect.
   
   // Calculate mean instar abundances:
   vector<Type> lambda_imm = exp(log_lambda_alpha + log_lambda_instar);

   // Likelihood evaluation for immatures:
   for (int i = 0; i < n_imm; i++){ 
      Type eta = 0;
      for (int k = 0; k < n_instar; k++){
         eta += lambda_imm[k] * (pnorm(x_imm[i] + dx / 2, mu[k], sigma[k]) - 
                                 pnorm(x_imm[i] - dx / 2, mu[k], sigma[k]));
      }
      v -= dpois(f_imm[i], eta, true);
   }
   
   // Likelihood evaluation for matures:
   vector<Type> lambda_mat(n_instar + 1);
   
   vector<Type> scale = Type(1) / (Type(1) + exp(-logit_scale));
   
   lambda_mat[0] = 0;
   for (int k = 1; k < (n_instar+1); k++){
      lambda_mat[k] = (Type(1) / (Type(1) + exp(-logit_scale[k-1]))) * lambda_imm[k-1];
   }
   for (int i = 0; i < n_mat; i++){ 
      Type eta = 0;
      for (int k = 0; k < (n_instar+1); k++){
         eta += lambda_mat[k] * (pnorm(x_mat[i] + dx / 2, mu[k], sigma[k]) - 
                                 pnorm(x_mat[i] - dx / 2, mu[k], sigma[k]));
      }
      v -= dpois(f_mat[i], eta, true);
   }
   
   // Export results:
   REPORT(mu);
   REPORT(sigma);
   REPORT(log_lambda_instar);
   REPORT(logit_scale);
   REPORT(scale);
   REPORT(lambda_imm);
   REPORT(lambda_mat);
   
   return v;
}

